kotlin/reflect/src/main/java/com/squareup/moshi/kotlin/reflect/KotlinJsonAdapter.kt

```diff
@@ -284,6 +372,333 @@ class KotlinJsonAdapterFactory : JsonAdapter.Factory {
 
     val nonTransientBindings = bindings.filterNotNull()
     val options = JsonReader.Options.of(*nonTransientBindings.map { it.name }.toTypedArray())
-    return KotlinJsonAdapter(constructor, bindings, nonTransientBindings, options).nullSafe()
+    return KotlinJsonAdapter(constructorData, bindings, nonTransientBindings, options).nullSafe()
+  }
+
+  private infix fun KmType?.isEqualTo(other: KmType?): Boolean {
+    return when {
+      this === other -> true
+      this != null && other != null -> {
+        // Note we don't check abbreviatedType because typealiases and their backing types are equal
+        // for our purposes.
+        arguments areEqualTo other.arguments
+            && classifier == other.classifier
+            && flags == other.flags
+            && flexibleTypeUpperBound isEqualTo other.flexibleTypeUpperBound
+            && outerType isEqualTo other.outerType
+      }
+      else -> false
+    }
+  }
+
+  private infix fun Collection<KmTypeProjection>.areEqualTo(
+      other: Collection<KmTypeProjection>): Boolean {
+    // check collections aren't same
+    if (this !== other) {
+      // fast check of sizes
+      if (this.size != other.size) return false
+      val areNotEqual = this.asSequence()
+          .zip(other.asSequence())
+          // check this and other contains same elements at position
+          .map { (fromThis, fromOther) -> fromThis isEqualTo fromOther }
+          // searching for first negative answer
+          .contains(false)
+      if (areNotEqual) return false
+    }
+    // collections are same or they are contains same elements with same order
+    return true
+  }
+
+  private infix fun KmTypeProjection?.isEqualTo(other: KmTypeProjection?): Boolean {
```

Comment:
yeah they explicitly won't use data classes or implement these because they wanted the classes to be as lightweight as possible for embedded envs ¯\_(ツ)_/¯